Add focal point support for images

DanielBiegler opened the pull request:
## Demo

Theres now focal point fields for files and a new mode inside the image editor which lets you pick a coordinate (the center of your selection). Then the asset service crops around that focal point instead of the center.

| New Tool | New Tool |
|--------|--------|
| ![Screenshot from 2023-12-21 12-29-43](https://github.com/directus/directus/assets/14810858/5713b24b-df69-4601-9b18-cce0bf0474c9) | ![Screenshot from 2023-12-21 12-29-48](https://github.com/directus/directus/assets/14810858/24d1e710-ed34-449f-bd39-f9fb7ea8cd0b) | 

| Cropping Mode | Focal Point Mode |
|--------|--------|
| ![Screenshot from 2023-12-21 12-30-18](https://github.com/directus/directus/assets/14810858/0d95aad4-5b23-4338-9ce1-1609e0f7824b) | ![Screenshot from 2023-12-21 12-30-13](https://github.com/directus/directus/assets/14810858/58a49b90-465c-4b71-a0d5-287fb08f5711) |

| Original | No focal point | With focal point on human |
|--------|--------|--------|
| ![image](https://github.com/directus/directus/assets/14810858/bcfbd14c-5efb-45b8-9a76-da6dda9e596e) | ![image](https://github.com/directus/directus/assets/14810858/939b93b7-9931-4dd5-919f-7110042b4152) | ![image](https://github.com/directus/directus/assets/14810858/182e5516-ee19-4208-a06b-7d62f1d7c544) | 

1. (Original photo by Photo by Juan Gomez from [Pexels](https://www.pexels.com/photo/photo-of-person-walking-alone-on-seashore-3002071/))
2. Example API requests with `?width=300&height=150`

New fields in the file detail view:

![image](https://github.com/directus/directus/assets/14810858/a884bb67-c529-4d09-ae96-defa037c0e64)

## Scope

What's changed:

- New focal point mode inside the image editor with custom styling
- New `focal_point_x`, `focal_point_y` columns in `directus_files` table that holds focal point coordinates
- New possible `focal_point_x`, `focal_point_y` asset transformation REST parameters for either using or overwriting an existing focal point
- Added to File-Type and Asset Transformations Type both in API and SDK
- Asset service will include cropping (sharp calls it extracting) if there are width & height parameters for the transformation when `fit=cover`
- Couple new translations
- Updated tests

## Potential Risks / Drawbacks

- If not done right we might transform images incorrectly which is of course - _not good_. I did preserve regular behaviour for all other modes since focal point cropping only makes sense for `cover`. Also made sure to fallback to regular sharp behaviour of scaling when resizing with only one dimension.

## Review Notes / Questions

- We could also add a way to set the focal point via the API. 
- Might want to include a section about this in the Docs
- Here are a couple of free test images with points of interest not in the center, so you can try out putting the focal point on them:
    - Vertical: https://www.pexels.com/photo/four-colourful-houses-2501965/
    - Vertical: https://www.pexels.com/photo/house-in-the-middle-of-crop-field-3330118/
    - Horizontal: https://www.pexels.com/photo/low-angle-photography-of-four-flying-birds-937766/
    - Horizontal: https://www.pexels.com/photo/white-clouds-2147451/
    - Square: https://www.pexels.com/photo/two-brown-cows-near-green-grass-fields-923556/

---

Fixes https://github.com/directus/directus/issues/20838

DanielBiegler committed:
add visual feedback for invalid value

DanielBiegler committed:
add focal point MVP

DanielBiegler committed:
Revert "add visual feedback for invalid value"

This reverts commit 1df186834273261ac1a6f0dea40be7e246e2601f.

Accidently committed some local testing stuff. Pls disregard! :)

DanielBiegler committed:
fix wrong cropping

DanielBiegler committed:
fix text for new cropping, import correct type

DanielBiegler committed:
fix saving

DanielBiegler committed:
place initial focal point to saved value or center, display different cancel text

DanielBiegler committed:
split up tooltips

DanielBiegler committed:
honor rotations & flips when saving focal point

DanielBiegler committed:
apply custom cropper styles for focal mode

DanielBiegler committed:
Create loud-crews-fix.md

DanielBiegler committed:
Merge remote-tracking branch 'origin/main' into add-focalpoint-cropping

DanielBiegler committed:
Merge remote-tracking branch 'refs/remotes/origin/add-focalpoint-cropping' into add-focalpoint-cropping

DanielBiegler committed:
add test and only crop when covering with fixed dimensions to preserve default behaviour

DanielBiegler committed:
Merge remote-tracking branch 'origin/main' into add-focalpoint-cropping

DanielBiegler committed:
linter gods pls forgive me

rijkvanzanten commented on api/src/database/migrations/20231215A-add-focalpoints.ts:
```diff
@@ -0,0 +1,13 @@
+
+export async function up(knex: Knex): Promise<void> {
+	await knex.schema.alterTable('directus_files', (table) => {
+		table.json('focal_point');
```
I'm wondering if we should normalize this to `focal_point_x` and `focal_point_y` integer columns

DanielBiegler replied:
My thought process was that the focal point infos are  _probably_ only needed all together. When would someone only fetch an x-coord and not the y-coord too? But not a hard requirement, easy change. Other opinions? :man_shrugging:

rijkvanzanten replied:
It's more about database modeling and normalization than use case for the data. Storing nested JSON with a known structure isn't very sql-like, as it effectively locks away the data in a non-structured blob.

rijkvanzanten commented on app/src/views/private/components/image-editor.vue:
```diff
@@ -147,6 +149,19 @@ function useImage() {
+			const x = (data?.x ?? 0) + (data?.width ?? 0) / 2;
+			const y = (data?.y ?? 0) + (data?.height ?? 0) / 2;
+			// Note: Required to include other fields before the `file` key
+			formData.append('focal_point', JSON.stringify({ x, y } satisfies File['focal_point']));
```
Should we 'just' do this as a regular JSON request? Right now we're technically re-uploading the whole file just to set the `focal_point` database column

DanielBiegler replied:
Yes we did it like that in the past too and I preserved this behaviour. I assume we did that so that we apply transformations like rotations and mirroring without needing to explicitly check for them(?). I guess it'd be also nice then to disable the saving button if there are no actual changes!

rijkvanzanten replied:
Rotation/mirror/crop etc all modify the original image, which is then replaced on disk. This is just metadata in the db though so that should go elsewhere 🙂

DanielBiegler replied:
Update: 

1. Save button now correctly gets disabled if there are no changes
2. Refactored persistence to collapse to one request if both the image and focal point change, but stay separate otherwise
3. Leave Image editor open if there are errors instead of closing

rijkvanzanten commented on api/src/utils/transformations.ts:
```diff
@@ -63,3 +98,71 @@ export function maybeExtractFormat(transforms: Transformation[]): string | undef
+	let width: number;
+	let height: number;
+
+	if (hRatio < wRatio) {
+		factor = hRatio;
+		height = Math.round(target.h);
+		width = Math.round(original.w / factor);
+	} else {
+		factor = wRatio;
+		width = Math.round(target.w);
+		height = Math.round(original.h / factor);
+	}
```
Will this work as expected for images where `hRatio == wRatio` ?

DanielBiegler replied:
Yes. Here are some debugging outputs for width:

```
api dev: --- 100x100 -> 98x100: { width: 100, height: 100, factor: 1 }
api dev: --- 100x100 -> 99x100: { width: 100, height: 100, factor: 1 }
api dev: --- 100x100 -> 100x100: { width: 100, height: 100, factor: 1 }
api dev: --- 100x100 -> 101x100: { width: 101, height: 101, factor: 0.9900990099009901 }
api dev: --- 100x100 -> 102x100: { width: 102, height: 102, factor: 0.9803921568627451 }
```

And for height:

```
api dev: --- 100x100 -> 100x98: { width: 100, height: 100, factor: 1 }
api dev: --- 100x100 -> 100x99: { width: 100, height: 100, factor: 1 }
api dev: --- 100x100 -> 100x100: { width: 100, height: 100, factor: 1 }
api dev: --- 100x100 -> 100x101: { width: 101, height: 101, factor: 0.9900990099009901 }
api dev: --- 100x100 -> 100x102: { width: 102, height: 102, factor: 0.9803921568627451 }
```

We can see that

1. If the target-dimensions are smaller than our original, the factor correctly doesnt change, since the target can be fully contained
2. If the target is equal to our original we also dont need to scale, factor stays
3. But if the target is larger, then we have to set the factor in order to scale the original so that we can fully cover the target

DanielBiegler committed:
replace json field with two int fields

DanielBiegler committed:
add focal point to sdk

DanielBiegler committed:
fix transformation for the two new db columns

DanielBiegler committed:
update test for new columns, add new tests

DanielBiegler committed:
wip: saving now differentiates between only img data and focal point and only enable saving if there are changes

but this is not optimal. would be better to check beforehand if we can collapse
to requests to one. Now its bad because
one request might succeed and the other fails.

DanielBiegler committed:
refactor image editor change persistence

now we save it in one request!

DanielBiegler committed:
Merge remote-tracking branch 'origin/main' into add-focalpoint-cropping

DanielBiegler committed:
Merge remote-tracking branch 'origin/main' into add-focalpoint-cropping

DanielBiegler committed:
Update loud-crews-fix.md

DanielBiegler committed:
add `focal_point_x` and `focal_point_y` to possible asset transformations

DanielBiegler committed:
Merge remote-tracking branch 'origin/main' into add-focalpoint-cropping

DanielBiegler committed:
Merge branch 'main' into add-focalpoint-cropping

br41nslug reviewed (commented):
What does the focal point selection "resizable circle range" do? Best i can gather is we do not actually save or use this value 🤔 
![firefox_G9PwjSEQGw](https://github.com/directus/directus/assets/9389634/c0aaf324-7bc9-45a3-ae23-a66eb1fcd0ca)

~~It also looks like the focal point does not get rendered anywhere after you have saved it 🤔 Is there a way to view the currently saved focal point in the App? I expected the image edit modal to be initialized with the previously saved focal point 🤔~~

DanielBiegler commented:
@br41nslug did you run the migration? The behaviour in your vid is broken for some reason.

The center x and y coordinate of said circle should get persisted to the db. Also reopening the focal point should also put it in the correct place. It does get initialized upon opening. 🤔

br41nslug commented:
Haha my bad I did forget that 😇 That does resolve the second question, however i do wonder if we should show a Dot where the focal point is in the editor without explicitly opening the focal point tool (it does render in the right place now after opening the focal tool in the image editor)

DanielBiegler committed:
fix assigning localdragmode upon cropper init

br41nslug commented on api/src/services/assets.ts:
```diff
@@ -65,7 +68,7 @@ export class AssetsService {
 		}
 
-		const file = (await this.knex.select('*').from('directus_files').where({ id }).first()) as File;
+		const file = (await this.filesService.readOne(id, { limit: 1 })) as File;
```
```suggestion
		const file = (await this.filesService.readOne(id, { limit: 1 })) as File | undefined;
```

Looks like the FilesService types are a bit whack (possibly inherited from ItemsService) we'll want to fix that at some point.

br41nslug commented on api/src/utils/transformations.ts:
```diff
@@ -63,3 +113,72 @@ export function maybeExtractFormat(transforms: Transformation[]): string | undef
+	target: Dimensions,
+	intermediate: Dimensions,
+): Region {
+	const newXCenter = focalPoint.x / factor;
+	const newYCenter = focalPoint.y / factor;
```
Do we need to check if `factor` is not `0` to prevent NaN's in the calculations? I think `clamp` will filter it out but does that affect the calculation done here?

br41nslug commented on app/src/views/private/components/image-editor.vue:
```diff
@@ -3,19 +3,18 @@ import api, { addTokenToURL } from '@/api';
-	width: number;
-	height: number;
-};
+type Image = Pick<
+	File,
+	'type' | 'filesize' | 'filename_download' | 'width' | 'height' | 'focal_point_x' | 'focal_point_y'
+>;
```
```suggestion
const imageFields = ['type', 'filesize', 'filename_download', 'width', 'height', 'focal_point_x', 'focal_point_y'] as (keyof File)[];

type Image = Pick<File, (typeof imageFields)[number]>;
```

DanielBiegler replied:
For me this leads to broken types.

1. It doesnt error if you include non-existing keys. So if `File` changes we wont be notified of errors.
2. It doesnt pick your selected keys, it picks all of the `File` keys, hence gives you completion for non-existing keys. We only want to fetch a subset.

Per 1)

![image](https://github.com/directus/directus/assets/14810858/56dec8ff-55c9-40ce-b30c-c3ec74299f5c)

Per 2)

![image](https://github.com/directus/directus/assets/14810858/fd04c017-7320-458b-b7c4-4245fa88b4f5)

Doing it explicitly gives us the correct type and reports back on errors for non-existing keys and will error in the future if `File` changes.

![image](https://github.com/directus/directus/assets/14810858/3c903bd1-8bdc-42a3-a664-c3eb8161a760)

br41nslug replied:
did i make a mistake in my type suggestion? 😅 It should be:
```ts
const imageFields = [
	'type',
	'filesize',
	'filename_download',
	'width',
	'height',
	'focal_point_x',
	'focal_point_y',
] as const;

type Image = Pick<
	File,
	typeof imageFields[number]
>;
```

![image](https://github.com/directus/directus/assets/9389634/44fcab72-a747-4660-8cda-1c764ce64081)
![image](https://github.com/directus/directus/assets/9389634/e7ae7d80-214e-463f-bfb6-daa877da65e4)

DanielBiegler replied:
Ahhhh nice! Thanks!

br41nslug commented on app/src/views/private/components/image-editor.vue:
```diff
@@ -132,7 +141,15 @@ function useImage() {
 			const response = await api.get(`/files/${props.id}`, {
 				params: {
-					fields: ['type', 'filesize', 'filename_download', 'width', 'height'],
+					fields: [
+						'type',
+						'filesize',
+						'filename_download',
+						'width',
+						'height',
+						'focal_point_x',
+						'focal_point_y',
+					] as (keyof Image)[],
```
```suggestion
					fields: imageFields,
```

paescuj commented on app/src/views/private/components/image-editor.vue:
```diff
@@ -375,19 +463,35 @@ function setAspectRatio() {
+		<div
+			v-if="imageData && !loading && !error"
+			class="editor-container"
+			:class="[localDragMode === 'focal_point' ? 'focal-point' : '']"
```
```suggestion
			:class="{ 'focal-point': localDragMode === 'focal_point' }"
```

DanielBiegler committed:
reuse fetched fields in type

Co-authored-by: Brainslug <br41nslug@users.noreply.github.com>

DanielBiegler committed:
update file type

Co-authored-by: Brainslug <br41nslug@users.noreply.github.com>

DanielBiegler committed:
update changeset

Co-authored-by: Pascal Jufer <pascal-jufer@bluewin.ch>

DanielBiegler commented on api/src/constants.ts:
```diff
@@ -38,7 +38,7 @@ export const SYSTEM_ASSET_ALLOW_LIST: TransformationParams[] = [
 ];
 
-export const ASSET_TRANSFORM_QUERY_KEYS: Array<keyof TransformationParams> = [
+export const ASSET_TRANSFORM_QUERY_KEYS = [
```
I removed it because it doesnt let us know whats actually inside of the array which then leads to wrong type inference down the line. I've ran into that specifically, thats why I removed it.

Let's say the `TransformationParams` changes in the future or we forget one key like so:

```ts
export const ASSET_TRANSFORM_QUERY_KEYS: Array<keyof TransformationParams> = [
	'key',
	'transforms',
	'width',
	'height',
	'format',
	'fit',
	'quality',
	'withoutEnlargement',
	'focal_point_x',
	// 'focal_point_y',
];
```

Now we dont know whether or not that key will ever be picked in the assets controller. The type doesnt let us know that we will never actually pick this key:

![image](https://github.com/directus/directus/assets/14810858/60492ead-a57c-41fc-81ba-ad194fddddfc)

By changing it to readonly we get the correct keys and avoid this point of failure.

Can we preserve the correct array items in the initialization but still mark it as readonly? :thinking: cc @br41nslug

DanielBiegler replied:
Talked to @br41nslug and he came up with the big brain solution of 

` as const satisfies (keyof TransformationParams)[]`

which does exactly what I wanted!!! :heart_eyes: 

With that we get the correct items, we error on wrong items and we do actually get the content of said array for the picking! win, win, win!

br41nslug replied:
Discussed with Daniel and this looks like the best of both worlds:
```ts
export const ASSET_TRANSFORM_QUERY_KEYS = [
	'key',
	'transforms',
	'width',
	'height',
	'format',
	'fit',
	'quality',
	'withoutEnlargement',
	'focal_point_x',
	'focal_point_y',
]  as const satisfies (keyof TransformationParams)[];
```

paescuj replied:
Right, this is the perfect solution 👍

DanielBiegler committed:
improve type for `ASSET_TRANSFORM_QUERY_KEYS`

Co-authored-by: Brainslug <br41nslug@users.noreply.github.com>

DanielBiegler committed:
Merge remote-tracking branch 'refs/remotes/origin/add-focalpoint-cropping' into add-focalpoint-cropping

DanielBiegler commented on api/src/utils/transformations.test.ts:
```diff
@@ -12,17 +12,21 @@ const inputFile: File = {
+	modified_on: '',
+	focal_point_x: null,
+	focal_point_y: null,
 };
```
That variable is already typed as `File` above:

```ts
const inputFile: File = {
```

Does that `satisfies` do something that I dont know? :thinking:

DanielBiegler replied:
Ah you meant it as combination with the other suggestions! Lemme see..

paescuj replied:
Yeah, it doesn't alter the type of `inputFile` (while still type checking) so when using the variable we know which props are available 👍

DanielBiegler committed:
Apply suggestions from code review

Trying out the batch change feature from github for the first time. Lets see.

Co-authored-by: Pascal Jufer <pascal-jufer@bluewin.ch>

DanielBiegler committed:
rename `persistChanges` to `saveImage`

DanielBiegler commented on app/src/views/private/components/image-editor.vue:
```diff
@@ -144,36 +161,88 @@ function useImage() {
 	}
 
-	function save() {
+	async function persistChanges(focalPoint?: { x: number | null; y: number | null }) {
```
Agreed!

notflip commented:
How can one go about testing this? Would love to try it out and provide feedback. Can I create a docker image from a pull request somehow?

paescuj commented:
> How can one go about testing this? Would love to try it out and provide feedback. Can I create a docker image from a pull request somehow?

Thanks for your interest!
You can either follow https://docs.directus.io/contributing/running-locally.html while checking out this branch with `git checkout add-focalpoint-cropping` after `git clone` or if you prefer to test via Docker you can run
```shell
git clone https://github.com/directus/directus.git
cd directus
git checkout add-focalpoint-cropping
docker build -t directus-dev .
```
and then follow https://docs.directus.io/self-hosted/quickstart.html while using your `directus-dev` image.

**Note:** I'm sure you are aware of this but mentioning just in case: Either of these methods are meant for testing/development purposes only and it's highly discouraged to run non-stable versions in productive environments.

phazonoverload committed:
Add docs for focal points (#20959)

* Add user guide

* Added to API Reference

* Prettier

* Spellchecker

* default null

Co-authored-by: Daniel Biegler <DanielBiegler@users.noreply.github.com>

* from -> around

Co-authored-by: Daniel Biegler <DanielBiegler@users.noreply.github.com>

* from -> around

---------

Co-authored-by: Daniel Biegler <DanielBiegler@users.noreply.github.com>

DanielBiegler committed:
add changeset for docs

DanielBiegler committed:
Merge remote-tracking branch 'origin/main' into add-focalpoint-cropping

DanielBiegler committed:
run prettier lets goooooooo

DanielBiegler committed:
move & show focal point fields and add divider

br41nslug reviewed (approved):
Checked the latest changes plus additions to the docs 😃

br41nslug committed:
Merge branch 'main' into add-focalpoint-cropping
